watchlogd-0.91a-5/watchlog.8
.TH WATCHLOG 8 "Mar 6 1997"
.UC 4
.SH NAME
watchlogd \-  periodically moves logs into a rotation before 
eventual deletion

.SH SYNOPSIS
\fBwatchlogd\fR [-DTV] [-a \fB<time>\fR] [-i \fB<interval>\fR] [-f 
\fB<statusfile>\fR] [\fB<configfile>\fR+]

\fBwatchlogd\fR [--debug] [--test] [--verbose]
                [--start-at \fB<time>\fR]
                [--interval \fB<interval>\fR] 
                [--state-file \fB<statusfile>\fR]
                [\fB<configfile>\fR+]

.SH DESCRIPTION
\fBwatchlogd\fR is a daemon that rotates backlogs daily, weekly,
monthly or
when a maximum file size is reached. It will wake up once or more a day 
and check if any logs need processing. Old logs can be saved for a number of
rotations and optionally compressed. The oldest log that is removed from
the rotational sequence can also be emailed before deletion. A test mode can
be enabled to debug and trace 
operational scenarios without actually processing log files. For singular
processing chores, the non-daemon variant \fBwatchlog\fR can be run.



.SH OPTIONS
.TP
\fB-T -\-test\fR
Turns on test mode which is used to 
observe potential behavior and will not actually modify logs
or the current operational state

.TP
\fB-V -\-verbose\fR
Turns on verbose mode which actually processes logs but outputs the same 
operational trace as test. If 
running in daemon mode then messages are sent to 
syslog including error messages which will also be sent out
if \fBsenderrors\fR is \fBon\fR. If running as \fBwatchlog\fR then 
messages go to stderr

.TP
\fB-D -\-debug\fR
Turns on debug mode which is the same as test mode except that deeper 
level messages are shown than in test mode.  
Debug mode is used to observe potential behavior and will not actually modify logs
or the current operational state.

.TP
\fB-a -\-start-at\fR \fItime\fR
If running in daemon mode, this is the time to start the initial
processing. \fItime\fR is a 24hr time (00:00 to 23:59).
Command line option overrides \fBstart\fR in \fB/etc/watchlogd.conf\fR

.TP
\fB-i -\-interval\fR \fIvalue\fR
If running in daemon mode, this is the time (in hours)
between each log check and processing.
Command line option overrides \fBinterval\fR in \fB/etc/watchlogd.conf\fR

.TP
\fB-s -\-state\-file <statusfile>\fR
Tells \fBwatchlog\fR to use an alternate operational state file.  This is 
useful if \fBwatchlog\fR is being run as different user(s) for various 
sets of logs.  The default state file is \fB/var/lib/watchlog/status\fR.

.TP
\fB-u -? -\-usage\fR
shows command line syntax

.TP
\fB-v -\-version\fR
shows version

.SH CONFIGURATION FILE

When run, \fBwatchlogd\fR reads 
\fB/etc/watchlog.conf\fR for runtime configuration. The 
parameters found before any specific 
logfile sections as treated as global settings. Some globals can be locally 
overriden in individual logfile config sections. 
\fBwatchlogd\fR will also attempt to parse config files located in the 
default config directory \fB/etc/watchlog.d\fR

Optionally, one or more config files (or directories of config files) may be 
given on the command line with options in the later files overriding
options in given in earlier files in the list.

Each
configuration file can set global options outside of a logfile 
section. each section for a particular logfile, can contain local settings 
that override the previous global ones

Example: /fB/etc/watchlog.conf/fR

.nf
.ta +3i

# /etc/watchlog.conf
# default config for watchlogd

# how to check log files: daily, weekly, monthly, filesize
checklog weekly

# keep backlogs for this number of rotations
rotations 4

# compression of backlogs?
compress on

# create a new log file after processing?
createnew on

# send removed log out?
sendlog on

# where to mail removed logs
sendlog admin@gizmonics.edu

# send out error log?
senderrors on

# where to send processing log
senderrorsto admin@gizmonics.edu


# include to use additional config files or directories
# specifiy how to handle wtmp and lastlog in this config file

/var/log/wtmp {

    checklog monthly
    rotations 12
    compress on
    sendlog off
    createnew on
    createmode 644 root root
}

/var/log/lastlog {

    checklog monthly
    rotations 12
    compress on
    sendlog off
    createnew on
    createmode 600 root root
}

.fi

.pp


.SH CONFIG FILE DIRECTIVES

.TP
\fBchecklog\fR {\fBdaily\fR|\fBweekly\fR|\fBmonthly\fR|\fBfilesize\fR}

check logs for possible rotation:

\fBdaily\fR Log files are rotation every day.

\fBweekly\fR Log files are rotated on the first day of the week.

\fBmonthly\fR Log files are rotated on the first day of every month.

\fBfilesize\fR Log files are rotated when \fBmaxsize\fR is reached

.TP
\fBmaxsize \fIbytes\fR
Log files are rotated when \fIbytes\fR size is reached if 
.\fBcheck\fR \fIfilesize\fR is specified. If
\fIbytes\fR is followed by \fIM\fR, then units are megabytes.
If suffix is \fIK\fR , then units are kilobytes. For example, \fBmaxsize 
 1048576\fR, \fImaxsize 1000K\fR, \fImaxsize 1M\fR are all equivalent.

.TP
\fBrotations \fIcount\fR
\fIcount\fR log files are kept in the rotation sequence before
being removed. the removed log can be mailed to the
address specified in the \fBsendlogto\fR directive before it
is deleted. If \fIcount\fR is 0 then no backlogs are saved in
the rotation sequence

.TP
\fBcompress\fR {\fBon\fR|\fBoff\fR}
If \fBon\fR, logs kept in the rotation sequence are compressed with \fBgzip\fR.

.TP
\fBcreatemode \fImode\fR [\fIowner\fR] [\fIgroup\fR]
If \fBcreatenew on\fR is specified, 
then after rotation (but before the \fBpostrotate\fR script is run)
a new log file is created (with the same name as the log file just rotated).
\fImode\fR specifies the mode for the log file in octal (the same
as \fBchmod(2)\fR), \fIowner\fR specifies the user name who will own the
log file, and \fIgroup\fR specifies the group the log file will belong
to 

.TP
\fBsendlog\fR {\fBon\fR|\fBoff\fR}
If \fBon\fR, last log removed from rotation is sent to
email address specified by \fBsendlogto\fR

.TP
\fBsendlogto \fIemailaddr\fR
Last log removed from rotation is sent to
\fIemailaddr\fR if \fBsendlog\fR is \fBon\fR

.TP
\fBsenderrors\fR {\fBon\fR|\fBoff\fR}
If \fBon\fR, a log of all error messages is
sent to email address specified by \fBsenderrorsto\fR

.TP
\fBsenderrorsto \fIemailaddr\fR
Any errors that occur during log file processing are mailed to 
\fIemailaddr\fR if \fBsenderrors\fR is \fBon\fR


.TP
\fBinclude\fB {\fIconfigdir\fR|\fIconfigfile\fR}
Reads the \fIconfigfile\fR file or the files in \fIconfigdir\fR
for config info.


.TP
\fBbegin postrotate\fR/\fBend\fR
The lines between \fIbegin postrotate\fR and \fIend\fR (both of which
must appear on lines by themselves) are executed after the log file is
rotated. These directives are only valid inside of a log 
file section. See \fBprerotate\fR as well.

.TP
\fBbegin prerotate\fR/\fBend\fR
The lines between \fBbegin prerotate\fR and \fBend\fR (both of which
must appear on lines by themselves) are executed before the log file is
rotated. These directives are only valid inside of a log 
file section. See \fBpostrotate\fR as well.


.SH FILES
.PD 0
.TP 27
\fI/etc/watchlogd.conf /var/lib/watchlogd/status\fR

.SH SEE ALSO
.IR gzip (1)

.SH BUGS
some

.SH AUTHOR
.nf
Mike Knerr <mk@eonova.com>

Many thanks to Erik Troan <ewt@redhat.com> for logrotate 
.fi


watchlogd-0.91a-5/CHANGES
watchlogd change log
--------------------

mike knerr <mk@enonova.com>

0.91 --> 0.91a

03/17/98
        * watchlog.c call to pidFetch does not log an error if pid file does
	  not exist since this function is also called when
	  verifying that the daemon is not already running
	  (ie. before daemon runs and if there is no pid file 
	  at that point then this is not an error)

        * log.c does not use a /tmp file for error log trace (for the usual reasons)
	  create a file in /var/lib/watchlogd instead 
	  
	* move unlink to remove error log file
	  thanks to Claus-Justus Heine <claus@momo.math.rwth-aachen.de>
	  for pointing it out
	    
	
0.00 --> 0.91 

03/06/97
        * log module will not print *any* msg on
          stdout or stderr if in daemon mode and not test mode 
          instead use verbose mode to get trace in syslog
          and/or set 'mailerrors on' to get mailed a 
          log of error level messages. if in daemon mode and
          test mode then some errorlogfile type errors will
          go to stderr for debuging purposes

        * recognize other exec pathnames like:
          /usr/sbin/watchlogd ./watchlogd

        * update man page

03/05/97 
        * add senderrors {on|off} option for
          global config file /etc/watchlog.conf

        * move errorlogfile handling into log module
          automatically opens errorfile when receives
          first err log msg. logSendErrorLog closes
          file out and attempts to send. errorlogfile
          is opened again on next err log msg

        * verbose for a daemon
          outputs test level log messages to syslog
          at info level and dups error level messages
          to errorlog and prints nothing to stderr. 
          using verbose mode with watchlog  is same
          as test mode

03/04/97
        * config is now loaded before waiting to start
          initial processing in anticipation of other
          global options in /etc/watchlogd.conf for
          start time and intervals

        * check is now checklog and has an off option in 
          anticipation of scanlog {on|off} 

        * verbose (-V) mode now shows same output as test mode
          but operations are actually executed

03/03/97
        * fix parsing of createmode options:

           - use strncpy into known buffer for sscanf for safety
         
           - detect and report the following errors:

               failed sscanf parse!!
               no args
               too many args
               invalid mode
               invalid owner
               invalid group
               valid mode
               valid mode,owner
               valid mode,owner,group
   
           - test mode shows createmode options set (if valid)

        * clean up test mode output for better readability

        * change createmode for /var/log/wtmp (644) and
          /var/log/lastlog (600) in default /etc/watchlog.conf

        * add some basic syslog logging

        * daemon periodically verifies saved pid and shuts down
          if not correct

        * command flags for message mode are all upper case
          (-D -T for debug and test) and debug mode is now available

03/02/97


        * add pid functions used to handle process pid and
          check for previously running daemon

        * dont run under an unrecognized name ie. not watchlog or 
          watchlogd

        * usage display adjusted for both variants watchlog and watchlogd

        * watchlogd detects itself and runs as daemon (without -d flag)

        * change messages levels
          test is now debug and debug is one level lower
          test is used to display trace for operating scenario
          and debug will be used to show info related more to
          code and system issues than to operation.



03/01/97

        * add sendlog {on|off} directive so can locally override
          whether to send out a log but keep a global email addr
          to send other logs to. for example, /var/log/wtmp should
          not be sent since it is binary (not ascii) but will want
          to retain send out removed log behavior from
          /etc/watchlog.conf

          new directive semantics for email addresses are:
          
           sendlogto <emailaddr>

           senderrorsto <emailaddr>

        * test mode shows postrotate script that would be executed
        
        * if in test mode show debug messages on stderr even if a
          mailing address is given for the processing log
 
        * directive to use a max file size to process is 'rotate filesize'
          and maxsize is used to set size in K or M and mazsize is
          not used if rotate is on a calendar interval like: daily,
          weekly, or monthly

        * don't allow processing interval > 24 hrs (must run at least
          once a day since checks are made for 1st of month and 
          1st of week )

        * don't reread config files especially if in daemon mode!


02/28/97

        * add optional parameters for daemon: processing interval and 
          first start time -- if not given ==> start time is now and 
          interval is 24 hrs  
        
        * add daemon mode

	* change command line option characters for --debug and --verbose

	* add option -v --version


02/26/97

	* clean up some syntax of logrotate.c code to
          be more semantically correct

	* add code to try default config file /etc/watchlog.conf
          and or directory /etc/watchlog.d if none specified on command 
          line (which is now optional)

	* change syntax of config file directives to be a more
          meaningful grammar 




from erik troan's logrotate

2.0.2 -> 2.1:
	- Don't output state information for logs that have never been
	  rotated (better then 1900-1-0)
	- Accept 1900-1-0 as time 0
	
2.0.1 -> 2.0.2:
	- I have no idea :-(

2.0 -> 2.0.1:
	- ignore files in included directories which end with ~, .rpmorig, or
	  .rpmsave
	  


watchlogd-0.91a-5/watchlog.c
/*
   watchlog.c - main code module for watchlogd log processing daemon.

   Copyright (c) 1997  mike knerr <mk@metrotron.com>

   This file is part of the watchlogd package

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program in the file COPYING; if not write to
   the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 
   02139, USA
 */

#include <alloca.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/types.h>
#include <signal.h>
#include <time.h>
#include <unistd.h>
#include <syslog.h>

#include "log.h"
#include "logrotate.h"
#include "watchlog.h"

#define PIDFILE "/var/run/watchlogd.pid"

/* private: */

static BOOLEAN bDaemon = FALSE;
static BOOLEAN bTestMode = FALSE;
static BOOLEAN bVerboseMode = FALSE;

static void usage(char *);
static void version(void);
static BOOLEAN isMyName(char *);
static BOOLEAN isMyDaemonName(char *);
static BOOLEAN pidSave(void);
static int pidFetch(void);
static BOOLEAN pidIsRunning(void);
static BOOLEAN pidIsMeRunning(void);
static BOOLEAN pidErase(void);
static BOOLEAN isValidStartTime(char *);
static BOOLEAN isTimeToStart(char *);
static void shutdown(int);


BOOLEAN inDaemonMode(void)
{
    return bDaemon;
}

BOOLEAN inTestMode(void)
{
    return bTestMode;
}

BOOLEAN inVerboseMode(void)
{
    return bVerboseMode;
}


static void usage(char *szExecName)
{

    if (strncmp(szExecName, "watchlogd", 9) == 0) {

	fprintf(stderr, "usage: watchlogd [-DTuv] [-s {<hh:mm>|now}] [-i <hours>] [-f <statefile>]\n"
		"                 [<configfile>+]\n");


    } else if (strncmp(szExecName, "watchlog", 8) == 0) {

	fprintf(stderr, "usage: watchlog [-DTuv] [-s {<hh:mm>|now}] [-i <hours>] [-f <statefile>]\n"
		"                 [<configfile>+]\n");


    } else {
	/* not called by us -- so get outa here */
	exit(1);
    }

    fprintf(stderr, "\n%s " VERSION "-" PATCHLEVEL "  " COPYRIGHT "\n", szExecName);
    fprintf(stderr, "May be freely redistributed under the"
	    " terms of the GNU Public License\n\n");

    exit(1);
}

static void version(void)
{
    fprintf(stderr, "watchlogd " VERSION "-" PATCHLEVEL "\n");
    exit(1);
}

static BOOLEAN isMyDaemonName(char *szName)
{

    int iReturn = FALSE;

    if (strncmp(szName, "watchlogd", 9) == 0) {
	iReturn = TRUE;
    } else if (strncmp(szName, "/usr/sbin/watchlogd", 19) == 0) {
	iReturn = TRUE;
    } else if (strncmp(szName, "./watchlogd", 11) == 0) {
	iReturn = TRUE;
    }
    return iReturn;
}


static BOOLEAN isMyName(char *szName)
{

    int iReturn = FALSE;

    if (isMyDaemonName(szName)) {
	return TRUE;
    } else if (strncmp(szName, "watchlog", 8) == 0) {
	iReturn = TRUE;
    } else if (strncmp(szName, "/usr/sbin/watchlog", 18) == 0) {
	iReturn = TRUE;
    } else if (strncmp(szName, "./watchlog", 10) == 0) {
	iReturn = TRUE;
    }
    return iReturn;
}


static BOOLEAN pidSave(void)
{

    FILE *f;
    int fd;
    int pid;

    if (((fd = open(PIDFILE, O_RDWR | O_CREAT, 0644)) == -1)
	|| ((f = fdopen(fd, "r+")) == NULL)) {

	message(MSG_ERR, "cannot open or create %s\n", PIDFILE);
	return FALSE;
    }
    if (flock(fd, LOCK_EX | LOCK_NB) == -1) {
	message(MSG_ERR, "cannot lock %s\n", PIDFILE);
	fclose(f);
	return FALSE;
    }
    pid = getpid();

    message(MSG_DEBUG, "writing out pid %d\n", pid);

    if (!fprintf(f, "%d\n", pid)) {
	message(MSG_ERR, "cannot write to %s (%s)\n", PIDFILE, strerror(errno));
	close(fd);
	return FALSE;
    }
    fflush(f);

    if (flock(fd, LOCK_UN) == -1) {
	message(MSG_ERR, "cannot unlock %s (%s)\n", PIDFILE, strerror(errno));
	close(fd);
	return FALSE;
    }
    close(fd);

    return TRUE;
}


static int pidFetch(void)
{

/* either return the value in the pidfile or return 0 if an error */

    FILE *f;
    int pid;

    if (!(f = fopen(PIDFILE, "r"))) {
	message(MSG_DEBUG, "cannot open %s\n", PIDFILE);
	return 0;
    }
    fscanf(f, "%d", &pid);
    fclose(f);

    message(MSG_DEBUG, "%s contains [%d] : my pid [%d]\n", PIDFILE, pid, getpid());

    return pid;
}


static BOOLEAN pidErase(void)
{

    if (unlink(PIDFILE) == 0) {
	return TRUE;
    }
    return FALSE;
}


static BOOLEAN pidIsMeRunning(void)
{

    int pid = pidFetch();

    /* see if my pidfile is me? */

    if (pid == getpid()) {
	message(MSG_DEBUG, "pid [%d] is me!\n", pid);
	return TRUE;
    }
    message(MSG_DEBUG, "pid [%d] is NOT me\n", pid);

    return FALSE;
}


static BOOLEAN pidIsRunning(void)
{

/* returns TRUE if there is general notion
   of the saved pid representing a running process */

    int pid = pidFetch();

    /* pidfile is gone or ... ? */

    if (pid == 0) {
	return FALSE;
    }
    /* is saved pid me */

    if (pidIsMeRunning()) {
	return TRUE;
    }
    /* DONT accidentally kill off every process or all in process group
     * attempting to do this test with pid <= -1 
     * in case pidfile was hacked
     */

    if (pid > 0) {
	if ((kill(pid, 0) == -1) && (errno == ESRCH)) {
	    message(MSG_DEBUG, "fetched pid [%d] is not a running process\n", pid);
	    return FALSE;
	} else {
	    message(MSG_DEBUG, "fetched pid [%d] is a running process\n", pid);
	}
    }
    return TRUE;
}



static void shutdown(int i)
{
    message(MSG_TEST, "Shutting down...\n");

    if (bDaemon) {

	/* do clean up */

	pidErase();
    }
    logSyslog(LOG_NOTICE, "Shutting down...\n");

    exit(i);
}



static BOOLEAN isValidStartTime(char *szTime)
{

    int iReturn = TRUE;

/* now is a valid start time! */

    if (strncmp(szTime, "now", 3) == 0) {
	return iReturn;
    }
    if (strlen(szTime) != 5) {
	iReturn = FALSE;
    }
    if (szTime[2] != ':') {
	iReturn = FALSE;
    }
    if (
	   (!isdigit(szTime[0])) ||
	   (!isdigit(szTime[1])) ||
	   (!isdigit(szTime[3])) ||
	   (!isdigit(szTime[4]))
	) {
	iReturn = FALSE;
    }
    if (
	   (szTime[0] > '2') ||
	   (szTime[3] > '5')
	) {
	iReturn = FALSE;
    }
    if (
	   (szTime[0] == '2') &&
	   (szTime[1] > '3')
	) {
	iReturn = FALSE;
    }
    if (iReturn == FALSE) {

	message(MSG_ERR, "Invalid initial processing time\n");
    }
    return iReturn;
}


static BOOLEAN isValidInterval(int i)
{

    if (i < 1) {
	message(MSG_ERR, "Invalid interval %d: Cannot run in less than hourly intervals\n", i);
	return FALSE;
    }
    if (i > 24) {
	message(MSG_ERR, "Invalid interval %d: Must run at least once a day\n", i);
	return FALSE;
    }
    return TRUE;
}


static BOOLEAN isTimeToStart(char *szT)
{

    time_t tNow;
    struct tm tmNow;
    char t[6];
    int iHour;
    int iMin;

/* check for valid start time anyway */

    if (isValidStartTime(szT) == FALSE) {

	message(MSG_ERR, "isTimeToStart got invalid start time\n");
	return FALSE;
    }
/* if starting immediately then don't wait */

    if (strncmp(szT, "now", 3) == 0) {
	message(MSG_TEST, "Initial starting time is now!\n");
	return TRUE;
    }
    tNow = time(NULL);
    tmNow = *localtime(&tNow);

    strncpy(t, szT, 5);
    t[5] = '\0';

    iHour = atoi(strtok(t, ":"));
    iMin = atoi(strtok(NULL, ":"));

    message(MSG_TEST, "check if time to start...\n");
    message(MSG_TEST, "time now: %d:%d \n", tmNow.tm_hour, tmNow.tm_min);
    message(MSG_TEST, "time to start: %d:%d \n", iHour, iMin);

    if ((iHour == tmNow.tm_hour) && (iMin == tmNow.tm_min)) {
	message(MSG_TEST, "ok to start\n");
	return TRUE;
    }
    message(MSG_TEST, "not yet...\n");
    return FALSE;
}



int main(int argc, char **argv)
{

    /* pass these defaults some of which will be overriden during parse */

    logInfo defConfig =
    {
	NULL,			/* */
	ROT_SIZE,		/* checklog filesize */
	1024 * 1024,		/* size is 1MB */
	0,			/* */
	0,			/* */
	NULL,			/* */
	NULL,			/* */
	NULL,			/* */
	0,			/* */
	NO_MODE,		/* createMode */
	NO_UID,			/* createUid  */
	NO_GID,			/* createGid  */
	FALSE,			/* mail removed log */
	FALSE,			/* mail error log */
	TRUE,			/* bCheckLog */
	FALSE			/* bScanLog */
    };


    int numLogs = 0, numStates = 0;
    logInfo *logs = NULL;
    logState *states = NULL;

    char *stateFile = DEFAULT_STATEFILE;
    char *szDefaultConfigFile = DEFAULTCONFFILE;
    char *szDefaultConfigDir = DEFAULTCONFDIR;

    int iInterval = 24;		/* interval to run in hours */
    int bDidFirstProcessing = FALSE;	/* so no reread config */
    char *szStartTime = "now";	/* time to start first processing */

    int i;
    int iHr;			/* hour counter */
    int rc = 0;
    int iOption, long_index;

/* will add:
 * -c check interval (for rotation)
 * -s scan interval (for events)
 * (both override -i)
 * -r {on|off} process rotations?
 * -e {on|off} scan events?
 */

    struct option options[] =
    {
	{"daemon", 0, 0, 'd'},
	{"debug-mode", 0, 0, 'D'},
	{"interval", 1, 0, 'i'},
	{"state-file", 1, 0, 'f'},
	{"start-at", 1, 0, 'a'},
	{"test-mode", 0, 0, 'T'},
	{"usage", 0, 0, '?'},
	{"verbose-mode", 0, 0, 'V'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
    };


    /* dont like the idea of not running under one of our names */

    if (!isMyName(argv[0])) {
	message(MSG_ERR, "Imposter! %s\n", argv[0]);
	exit(1);
    }
    logSetLevel(MSG_NORMAL);

    if (isMyDaemonName(argv[0])) {

	logSetMyName("watchlogd");
    } else {
	logSetMyName("watchlog");
    }

    /* get parameters and set flags */

    while (1) {
	iOption = getopt_long(argc, argv, "DTVdf:i:a:uv?", options, &long_index);
	if (iOption == EOF)
	    break;

	switch ((char) iOption) {

	case 'f':
	    stateFile = optarg;
	    break;

	case 'd':
	    bDaemon = TRUE;
	    break;

	case 'D':
	    bTestMode = TRUE;
	    logSetLevel(MSG_DEBUG);
	    break;

	case 'i':
	    iInterval = atoi(optarg);
	    break;

	case 'a':
	    szStartTime = optarg;
	    break;

	case 'T':
	    bTestMode = TRUE;
	    logSetLevel(MSG_TEST);
	    break;

	case 'u':
	    usage(argv[0]);
	    break;

	case 'V':
	    logSetLevel(MSG_TEST);
	    bTestMode = FALSE;
	    bVerboseMode = TRUE;
	    break;

	case 'v':
	    version();
	    break;

	case '?':
	    usage(argv[0]);
	    break;
	}
    }


/* see if started as watchlogd 
 * and if so then run as daemon
 */

    if (isMyDaemonName(argv[0])) {

	bDaemon = TRUE;

    }
/* check out validity of daemon option parameters before continuing */

    /* if we are daemon may have an optional start time or interval */

    if (bDaemon == TRUE) {

	message(MSG_TEST, "Daemon mode is on\n");
	message(MSG_TEST, "Processing interval is: %d\n", iInterval);


	if (isValidInterval(iInterval) == FALSE) {
	    exit(1);
	}
	if (isValidStartTime(szStartTime) == FALSE) {
	    exit(1);
	}
	message(MSG_TEST, "Initial processing time: %s\n", szStartTime);


    }				/*endif is daemon */
    /* if we will be a daemon then fork now */
    if (bDaemon) {

	if (!pidIsRunning()) {

	    message(MSG_TEST, "Starting watchlog daemon\n");

	    switch (fork()) {

	    case -1:
		message(MSG_ERR, "Error forking daemon [%d]\n", getpid());
		exit(0);
		break;

	    case 0:
		/* this process is the child
		 * so continue to run
		 */
		message(MSG_TEST, "daemon fork is OK! [%d]\n", getpid());
		(void) setsid();

		logSyslog(LOG_NOTICE, "Starting Up... (watchlogd-" VERSION
			  "-" PATCHLEVEL ")\n");
		if (!pidSave()) {
		    shutdown(1);
		}
		break;

	    default:
		/* this process is the parent process
		 * so exit thereby letting only the 
		 * child continue to run as the daemon
		 */

		message(MSG_TEST, "daemon parent is exiting [%d]\n", getpid());
		_exit(0);
	    }

	} else {

	    message(MSG_ERR, "Daemon is already running!\n");
	    logSyslog(LOG_NOTICE, "Exiting... Daemon is already running!\n");

	    exit(1);

	}

    }				/*endif if daemon mode */
/****** read log configurations *******/
    /*  check to see if a config file(s) was given */
    if (optind == argc) {

	/* if no config file given then attempt to use default config file */

	message(MSG_DEBUG, "szDefaultConfigFile: %s\n", szDefaultConfigFile);

	if (readConfigPath(szDefaultConfigFile, &defConfig, &logs, &numLogs)) {

	    /*
	     * if no default config file then attempt to use
	     * the default directory of config files
	     */

	    if (readConfigPath(szDefaultConfigDir, &defConfig, &logs, &numLogs)) {
		shutdown(1);
	    }
	}
    } else {

	/* otherwise a primary config file was given so attempt use it */

	while (optind < argc) {
	    if (readConfigPath(argv[optind], &defConfig, &logs, &numLogs)) {
		shutdown(1);
	    }
	    optind++;
	}
    }


    /* main server loop */
    while (1) {

	if (bDaemon) {

/* wait until time to start first processing */

	    if (bDidFirstProcessing == FALSE) {

		message(MSG_TEST, "Waiting to start initial processing at '%s'\n", szStartTime);
		sleep(1);

		/* hang out here until time to start initial processing */

		while (1) {

		    /* message(MSG_DEBUG, "Check start time in while(1) \n"); */

		    if (isTimeToStart(szStartTime)) {
			message(MSG_TEST, "Starting initial processing...\n");
			break;
		    }
		    sleep(60);
		}
	    }
	}			/* endif a daemon */
	if (readState(stateFile, &states, &numStates)) {
	    shutdown(1);
	}
	if (bDaemon) {
	    /* be a little paranoid */
	    if (!pidIsMeRunning()) {
		message(MSG_ERR, "Saved PID is NOT me!\n");
		logSyslog(LOG_ERR, "Cannot verify my PID!\n");
		shutdown(1);
	    }
	}
	message(MSG_TEST, "\nBegin checking %d logs\n", numLogs);
	logSyslog(LOG_INFO, "Processing logs...\n");

	for (i = 0; i < numLogs; i++) {
	    rc |= rotateLog(logs + i, &states, &numStates);
	}

	if (!bTestMode && writeState(stateFile, states, numStates)) {
	    shutdown(1);
	}
	bDidFirstProcessing = TRUE;


	if (bDaemon) {

	    /* now sleep for the interval until next time to run through */

	    message(MSG_TEST, "\nWaiting for %d hours to start next log processing\n", iInterval);

	    for (iHr = 1; iHr <= iInterval; iHr++) {

		/* can do stuff here every hour */

		if (!pidIsMeRunning()) {

		    message(MSG_ERR, "Saved PID is NOT me!\n");
		    logSyslog(LOG_ERR, "Cannot verify my PID!\n");
		    shutdown(1);
		}
		sleep(60 * 60);

	    }



	}
	if (!bDaemon) {
	    message(MSG_TEST, "\nNot in daemon mode so exiting now...\n");
	    break;
	}
    }				/*end while server loop */

    shutdown(rc != 0);

    return (0);			/* gcc squawks about this not being here */
}				/*end main */
